### PR TITLE
Upgrade model to GPT2 and dataset to tiiuae/falcon-refinedweb

### PR DESCRIPTION
**Pull request by KMFODA**
Todos:
- [x] Add gradient compression on miner and validator sides to reduce GPT2 `gradient_averaging` times
- [x] Add state compression on miner and validator sides to reduce GPT2 `load_state_from_paramater_peers`
- [x] Add a dataset.py file that streams and filters the `tiiuae/falcon-refinedweb` dataset
- [x] Convert the list of dataset indices we've trained on to a bitarray to be able to store a much larger list size
- [x] Add gradient checkpointing to remove OOM errors
- [x] Switch the DHT on the validator side to have client_mode=True
- [x] Replace the validator state_averager with a optimiser
- [x] Only perform local loss checks on miners when the global_epoch hasn't changed and therefore when the state hasn't been updated yet
- [x] Store`dataset_indices_test` in the DHT so that all validators can reference the same sub-dataset for testing and there have much more similar weights thereby improving vtrust issues.
- [ ] Update configs to main net before merge

Decommissioned to next release:
- [ ] Store `dataset_indices_train` in the DHT so that all validators can be aligned on which dataset_indices have been used for training yet vs not. Thereby enabling true Data Parallelism.

### CODE REVIEW

File: neurons/miner.py [hunk 1]
```diff
@@ -173,51 +177,58 @@ async def forward(
...
+                
+                if not self.config.neuron.dont_wandb_log:
+                    self.wandb.log({"loss": outputs.loss.detach().item(), "opt_local_epoch": self.opt.local_epoch})
+
+            torch.cuda.empty_cache()
 
-        average_loss = total_loss / len(dataloader)
-        synapse.loss = average_loss
+        synapse.loss = outputs.loss.detach().item()
+        synapse.epoch = self.opt.tracker.local_progress.epoch
```
[the remark this answers is not in the document]
Comment (KMFODA) [reply neurons/miner.py:229]: Yes exactly so the if statement is to try and detect if the state has changed in the time between the miner's training loop and the validator's validation loop. Because if it has then the validator will be running on a different state and we'll be penalising the miner for having a different loss to that of the validator just because of timing.

[the remark this answers is not in the document]
Comment (KMFODA) [reply neurons/miner.py:229]: The average loss was removed because I couldn't get it to work properly late night yesterday whereas normal loss was working fine. I can re-think wether we can add it with a fresh set of eyes today. The main issue will be how do we average a loss that is already averaged across gradient_accumilation_steps.

File: template/validator/validator_core.py [hunk 2]
```diff
@@ -110,3 +97,28 @@ def upload_checkpoint(commit_message, state_averager, model, repo_path, repo_url
         commit_message=commit_message,
     )
     bt.logging.info("Finished uploading to Model Hub")
+
+if __name__ == "__main__":
```
Comment (KMFODA) [inline template/validator/validator_core.py:101]: @mikkelmedm was this only supposed to be here for testing?

**Review by fredesere — APPROVED** [review]

**Review by fredesere — APPROVED** [review]

**Review by mekaneeky — APPROVED** [review]
